### PR TITLE
PLAT-995 Use Chrome nodes for Selenium tests

### PR DESCRIPTION
**Pull request by alexander-schmidt**
- We now execute Selenium tests on Chrome nodes, instead of Firefox nodes.
  - Eliminated incomplete implementation of multiplexed test execution.
  - Removed unnecessary dependencies.
- Eliminated `AjaxSeleniumLowLevelTestRetryingRunner`.
  - Blunt retying of all tests is generally unacceptable because it conceals nondeterminism.
  - Introduced `AjaxSeleniumLowLevelTestEngine.discardWebDriver` as a replacement, for explicit invocation in unit tests when neccessary.
- All viewport resizing operations now use `AjaxSeleniumLowLevelTestEngineViewport.setViewportSize`.
  - Now also called upon each invocation of `openTestNode`.
  - Fixed several glitches in `setViewportSize`, and made it wait until the viewport size has actually changed; compensates for a delay in Chrome.
- Fixed several issues with existing tests.
  - Added missing `waitForServer` calls across various Selenium low-level tests.
  - Fixed raciness in `AjaxDomFocusAndBlurEventTest`.
- Renamed `AjaxSeleniumTestSegment` to `AjaxSeleniumTestArea` (fixes a misnomer).

### CODE REVIEW

File: platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/grid/SleniumGridDockerComposeFileGenerator.java [hunk 1]
```diff
@@ -91,23 +90,21 @@ private void appendSeleniumHubService(OutputBuilder builder) {
...
-				builder.appendLine("      - %s", generateHubName());
-				builder.appendLine("    environment:");
-				builder.appendLine("      - SE_EVENT_BUS_HOST=%s", generateHubName());
-				builder.appendLine("      - SE_EVENT_BUS_PUBLISH_PORT=%s", hubConfiguration.getPortEventBusPublish());
-				builder.appendLine("      - SE_EVENT_BUS_SUBSCRIBE_PORT=%s", hubConfiguration.getPortEventBusSubscribe());
-				builder.appendLine("    networks:");
-				builder.appendLine("      - %s", generateNetworkName());
-				builder.appendLine("");
-			}
+		for (int nodeIndex = 1; nodeIndex <= calculateNodeCount(nodeConfiguration); nodeIndex++) {
```
Comment (oliver-richers) [inline platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/grid/SleniumGridDockerComposeFileGenerator.java:93]: should be extracted to variable
```suggestion
		var nodeCount =  calculateNodeCount(nodeConfiguration);
		for (int nodeIndex = 1; nodeIndex <= nodeCount; nodeIndex++) {
```

File: platform-ajax/src/test/java/com/softicar/platform/ajax/dom/event/click/AjaxDomClickEventTest.java [hunk 2]
```diff
@@ -65,6 +67,9 @@ public void testWithScrolledWindow() {
 		AjaxSeleniumTestSegment backgroundDivSize = getSize(backgroundDiv);
 		setViewportSize(backgroundDivSize.getWidth() - SCROLL_X, backgroundDivSize.getHeight() - SCROLL_Y);
 
+		// race condition: resizing apparently takes some time
+		Sleep.sleep(Duration.ofMillis(500));
```
Comment (oliver-richers) [inline platform-ajax/src/test/java/com/softicar/platform/ajax/dom/event/click/AjaxDomClickEventTest.java:71]: that is nasty, solutions?

**Review by oliver-richers — CHANGES_REQUESTED** [review]

File: platform-ajax/src/test/java/com/softicar/platform/ajax/dom/event/click/AjaxDomClickEventTest.java (hunk 2 above)
Comment (alexander-schmidt) [reply platform-ajax/src/test/java/com/softicar/platform/ajax/dom/event/click/AjaxDomClickEventTest.java]: Changed `AjaxSeleniumLowLevelTestEngineViewport.setViewportSize` to use a `WebDriverWait` with a timeout of 10s instead. Pls check it out.

File: platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/engine/level/low/AjaxSeleniumLowLevelTestEngineViewport.java [hunk 3]
```diff
@@ -34,7 +36,12 @@ public void setViewportSize(int width, int height) {
 		AjaxSeleniumTestSegment viewportSize = getViewportSize();
 		int extraWidth = windowSize.getWidth() - viewportSize.getWidth();
 		int extraHeight = windowSize.getHeight() - viewportSize.getHeight();
-		getWindow().setSize(new org.openqa.selenium.Dimension(width + extraWidth, height + extraHeight));
+
+		getWindow().setSize(new Dimension(width + extraWidth, height + extraHeight));
+
+		// wait for the window size to actually change
+		new WebDriverWait(webDriverSupplier.get(), Duration.ofSeconds(10))//
+			.until(drv -> !getViewportSize().equals(viewportSize));
```
Comment (oliver-richers) [inline platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/engine/level/low/AjaxSeleniumLowLevelTestEngineViewport.java:44]:
```suggestion
		new WebDriverWait(webDriverSupplier.get(), Duration.ofSeconds(10))//
			.until(driver -> !getViewportSize().equals(viewportSize));
```
not sure why we use `WebDriverWait` here and `FluentWait` everywhere else, what is the difference?

File: platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/grid/SleniumGridDockerComposeFileGenerator.java [hunk 4]
```diff
@@ -91,23 +90,22 @@ private void appendSeleniumHubService(OutputBuilder builder) {
...
-				builder.appendLine("      - %s", generateHubName());
-				builder.appendLine("    environment:");
-				builder.appendLine("      - SE_EVENT_BUS_HOST=%s", generateHubName());
-				builder.appendLine("      - SE_EVENT_BUS_PUBLISH_PORT=%s", hubConfiguration.getPortEventBusPublish());
-				builder.appendLine("      - SE_EVENT_BUS_SUBSCRIBE_PORT=%s", hubConfiguration.getPortEventBusSubscribe());
-				builder.appendLine("    networks:");
-				builder.appendLine("      - %s", generateNetworkName());
-				builder.appendLine("");
-			}
+		var nodeCount =  calculateNodeCount(nodeConfiguration);
```
Comment (oliver-richers) [inline platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/grid/SleniumGridDockerComposeFileGenerator.java:93]: auto-formatter broken?
```suggestion
		var nodeCount = calculateNodeCount(nodeConfiguration);
```

**Review by oliver-richers — CHANGES_REQUESTED** [review]

File: platform-ajax/src/test/java/com/softicar/platform/ajax/dom/event/click/AjaxDomClickEventTest.java (hunk 2 above)
Comment (oliver-richers) [reply platform-ajax/src/test/java/com/softicar/platform/ajax/dom/event/click/AjaxDomClickEventTest.java]: better, see comment

File: platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/grid/SleniumGridDockerComposeFileGenerator.java (hunk 4 above)
Comment (alexander-schmidt) [reply platform-ajax/src/main/java/com/softicar/platform/ajax/testing/selenium/grid/SleniumGridDockerComposeFileGenerator.java:93]: Well, I just committed the suggestion you made earlier ^^ ...and missed the double-space.

**Review by oliver-richers — APPROVED** [review]

**Pull request merged**